---
title: "QC & PCA for merged samples [Normalized]"
author: "ATIA Safiya"
date: "2023-07-26"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: lumen
---

# Seurat object  
## Load required packages
```{r,  message=FALSE, warning=FALSE}
library(Seurat)
library(tidyverse)
library(ggplot2)
library("scales")
```

## Load seurat objects (RAW)

```{r}
BT1.diff <- readRDS(file = './results/rds/samples/Norm_BT1_diff.rds')
BT1.prolif <- readRDS(file = './results/rds/samples/Norm_BT1_prolif.rds')

BT2.diff <- readRDS(file = './results/rds/samples/Norm_BT2_diff.rds')
BT2.prolif <- readRDS(file = './results/rds/samples/Norm_BT2_prolif.rds')

BT54.diff <- readRDS(file = './results/rds/samples/Norm_BT54_diff.rds')
BT54.prolif <- readRDS(file = './results/rds/samples/Norm_BT54_prolif.rds')

BT88.diff <- readRDS(file = './results/rds/samples/Norm_BT88_diff.rds')
BT88.prolif <- readRDS(file = './results/rds/samples/Norm_BT88_prolif.rds')

LGG85.diff <- readRDS(file = './results/rds/samples/Norm_LGG85_diff.rds')
LGG85.prolif <- readRDS(file = './results/rds/samples/Norm_LGG85_prolif.rds')

LGG275.diff <- readRDS(file = './results/rds/samples/Norm_LGG275_diff.rds')
LGG275.prolif <- readRDS(file = './results/rds/samples/Norm_LGG275_prolif.rds')

LGG336.diff <- readRDS(file = './results/rds/samples/Norm_LGG336_diff.rds')
LGG336.prolif <- readRDS(file = './results/rds/samples/Norm_LGG336_prolif.rds')

LGG349.diff <- readRDS(file = './results/rds/samples/Norm_LGG349_diff.rds')
LGG349.prolif <- readRDS(file = './results/rds/samples/Norm_LGG349_prolif.rds')
```

```{r}
seurat.list <-list(BT1.diff, BT1.prolif, BT2.diff, BT2.prolif, BT54.diff, BT54.prolif, BT88.diff, BT88.prolif,
                  LGG85.diff, LGG85.prolif, LGG275.diff, LGG275.prolif, LGG336.diff, LGG336.prolif, LGG349.diff, LGG349.prolif)
```

##  <span style="color:#296d98;"> Parameters </span>  
```{r}
pc = 60
```

# Integrate data
Pearson residuals are used for identifying variable features, PCA, dimensionality reduction and integration.
The corrected counts should be used for DE/visualization of marker genes.

```{r, warning=FALSE, message=FALSE}
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 3000)
seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = features)

anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT", anchor.features = features)
```

```{r}
seurat_obj <- IntegrateData(anchorset = anchors, normalization.method = "SCT")
DefaultAssay(seurat_obj) <- "integrated"

seurat_obj
table(seurat_obj$orig.ident)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
## delete those objects from memory
rm(BT1.diff, BT1.prolif, BT2.diff, BT2.prolif, BT54.diff, BT54.prolif, BT88.diff, BT88.prolif,
                  LGG85.diff, LGG85.prolif, LGG275.diff, LGG275.prolif, LGG336.diff, LGG336.prolif, LGG349.diff, LGG349.prolif)
```

```{r}
counts <- GetAssayData(seurat_obj, assay = "integrated")
print(length(rownames(counts)))

counts <- GetAssayData(seurat_obj, assay = "RNA")
print(length(rownames(counts)))

counts <- GetAssayData(seurat_obj, assay = "SCT")
print(length(rownames(counts)))
```

## Zero counts  
See features without any counts left after the filtering.

```{r}
counts <- GetAssayData(seurat_obj, assay = "SCT")
zero <- which(rowSums(counts) == 0 )
length(zero)
```

# QC metrics (after integration) 
```{r, message=FALSE, warning=FALSE}
metadata <- seurat_obj@meta.data
```

## Nb of reads  
```{r, message=FALSE}
VlnPlot(seurat_obj,features="nCount_RNA", group.by = "orig.ident", pt.size = 0)
RidgePlot(seurat_obj, features = "nCount_RNA", group.by = "orig.ident")

metadata %>% 
  ggplot(aes( x=nCount_RNA,y=orig.ident, fill= orig.ident)) + 
	geom_boxplot() + 
	theme_classic()+
  scale_x_log10()
```

## Nb of genes  
```{r, message=FALSE}
VlnPlot(seurat_obj,features="nFeature_RNA", group.by = "orig.ident", pt.size = 0)
RidgePlot(seurat_obj, features = "nFeature_RNA", group.by = "orig.ident")

metadata %>% 
  ggplot(aes( x=nFeature_RNA,y=orig.ident, fill= orig.ident)) + 
	geom_boxplot() + 
	theme_classic()+
  scale_x_log10()
    
```

```{r}
ggplot(seurat_obj[[]]) + 
   geom_histogram(aes(nFeature_RNA), 
                  color = "#558bdc", fill= "#173664",
                  binwidth = 50) + 
  ggtitle("Distribution of nFeature_RNA ") + NoLegend()
```

## Gene expression  
We can pick the first 100 cells and look at the distributions of their expression values.

```{r, warning=FALSE}
apply(seurat_obj@assays$SCT@data,1,mean) -> gene.expression
sort(gene.expression, decreasing = TRUE) -> gene.expression

as.tibble(
  seurat_obj@assays$SCT@data[,1:100]
) %>%
  pivot_longer(
    cols=everything(),
    names_to="cell",
    values_to="expression"
  ) %>%
  ggplot(aes(x=expression, group=cell)) +
  geom_density() +
  coord_cartesian(ylim=c(0,0.6), xlim=c(0,3))
```

## MT genes  
*calculating pct on RNA assay (raw)*
```{r, message=FALSE}
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-", assay = 'RNA')
VlnPlot(seurat_obj, features = "percent.mt", group.by = "orig.ident", pt.size = 0)
RidgePlot(seurat_obj, features = "percent.mt", group.by = "orig.ident")
```

```{r}
ggplot(seurat_obj[[]]) + 
   geom_histogram(aes(percent.mt), 
                  color = "#558bdc", fill= "#173664",
                  binwidth = 0.5) + 
  ggtitle("Distribution of Percentage Mitochondrion") + NoLegend()
```

## Ribosomal genes  
```{r, message=FALSE}
PercentageFeatureSet(seurat_obj,pattern="^RP[LS]", assay = 'RNA') -> seurat_obj[["percent.ribosomal"]] 
VlnPlot(seurat_obj, features = "percent.ribosomal", group.by = "orig.ident", pt.size = 0)
RidgePlot(seurat_obj, features = "percent.ribosomal", group.by = "orig.ident")
```

```{r}
ggplot(seurat_obj[[]]) + 
   geom_histogram(aes(percent.ribosomal), 
                  color = "#558bdc", fill= "#173664",
                  binwidth = 0.7) + 
  ggtitle("Distribution of Percentage Ribosomal") + NoLegend()
```

## Largest gene  
```{r, message=FALSE}
apply(
  seurat_obj@assays$SCT@counts,
  2,
  max
) -> largest_count

apply(
  seurat_obj@assays$SCT@counts,
  2,
  which.max
) -> largest_index

rownames(seurat_obj)[largest_index] -> seurat_obj$largest_gene

100 * largest_count / seurat_obj$nCount_RNA -> seurat_obj$percent.largest_gene

VlnPlot(seurat_obj,features="percent.largest_gene", group.by = "orig.ident", pt.size = 0)
RidgePlot(seurat_obj, features = "percent.largest_gene", group.by = "orig.ident")
```

```{r}
ggplot(seurat_obj[[]]) + 
   geom_histogram(aes(percent.largest_gene), 
                  color = "#558bdc", fill= "#173664",
                  binwidth = 0.7) + 
  ggtitle("Distribution of Percentage Largest Gene") + NoLegend()
```

## FeatureScatter  
```{r}
FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by ="orig.ident")
FeatureScatter(seurat_obj,feature1 = "nCount_RNA", feature2 = "percent.largest_gene",group.by ="orig.ident")

FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",group.by ="orig.ident") + geom_smooth(method = 'lm')
```

## Cell cycle scoring  
```{r, warning=FALSE}
seurat_obj <- CellCycleScoring(seurat_obj, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = TRUE)

seurat_obj@meta.data %>%
  group_by(orig.ident,Phase) %>%
  count() %>%
  group_by(orig.ident) %>%
  mutate(percent=100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=orig.ident,y=percent, fill=Phase)) +
  scale_fill_manual( values=c(G1='azure4',G2M='dodgerblue',S='tomato3')) +
  geom_col() +
  ggtitle("Percentage of cell cycle phases per sample")

as_tibble(seurat_obj[[]]) %>%
  ggplot(aes(Phase, fill=Phase)) +
  scale_fill_manual( values=c(G1='azure4',G2M='dodgerblue',S='tomato3'))+ geom_bar()


as_tibble(seurat_obj[[]]) %>%
  ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + 
  scale_color_manual( values=c(G1='azure4',G2M='dodgerblue',S='tomato3')) +
  geom_point() +
  coord_cartesian(xlim=c(-0.15,0.15), ylim=c(-0.15,0.15))
```

# Identify high variable features  
*Included in the `SCTransform` function.*

```{r, eval=FALSE}
# identify the 10 most highly variable genes
top10 <- head(VariableFeatures(seurat_obj), 10)

# plot the variable features 
plot <- VariableFeaturePlot(seurat_obj)
LabelPoints(plot = plot, points = head(top10,10), repel = TRUE)
```

We can plot out a graph of the variance vs mean and highlight the selected genes so we can see whether we think weâ€™re likely to capture what we need.    

```{r, eval=FALSE}
as_tibble(HVFInfo(seurat_obj),rownames = "Gene") -> variance.data

variance.data %>% 
  mutate(hypervariable=Gene %in% VariableFeatures(seurat_obj)
) -> variance.data

variance.data %>% 
  ggplot(aes(log(mean),log(variance),color=hypervariable)) + 
  geom_point() + 
  scale_color_manual(values=c("black","red"))
```

# Linear dimensional reduction  
## PCA  
```{r}
seurat_obj <- RunPCA(seurat_obj, verbose = FALSE)
VizDimLoadings(seurat_obj, dims = 1:2, reduction = "pca")
```

```{r}
mat <- Seurat::GetAssayData(seurat_obj, assay = "SCT", slot = "scale.data") ## instead of RNA
pca <- seurat_obj[["pca"]]

# Get the total variance:
total_variance <- sum(matrixStats::rowVars(mat))
eigValues = (pca@stdev)^2  ## EigenValues

varExplained = eigValues / total_variance

pc1 = percent(varExplained[1], accuracy = 0.01)
pc2 = percent(varExplained[2], accuracy = 0.01)
```

```{r}
DimPlot(seurat_obj, reduction = "pca", group.by = "orig.ident") + xlab(paste0("PC_1 : ", pc1)) + ylab(paste0("PC_2 :", pc2 ))
DimPlot(seurat_obj, reduction = "pca", group.by = "Phase", cols = c(G1='azure4',G2M='dodgerblue',S='tomato3'))

DimPlot(seurat_obj, reduction = "pca", group.by = "orig.ident", label.size = 4, label = TRUE)
```

### Split by condition
```{r}
DimPlot(seurat_obj, reduction = "pca", group.by="orig.ident", split.by = "condition")
DimPlot(seurat_obj, reduction = "pca", group.by = "orig.ident", split.by = "condition", label.size = 4, label = TRUE)+ NoLegend()
DimPlot(seurat_obj, reduction = "pca", group.by="orig.ident", split.by = "condition")+ NoLegend()
```

### Split by line
```{r, eval=FALSE}
DimPlot(seurat_obj, reduction = "pca", group.by="orig.ident", split.by = "orig.ident")
DimPlot(seurat_obj, reduction = "pca", group.by = "orig.ident", split.by = "orig.ident", label.size = 4, label = TRUE)+ NoLegend()
DimPlot(seurat_obj, reduction = "pca", group.by="orig.ident", split.by = "orig.ident")+ NoLegend()
```

## Save seurat object  
```{r}
saveRDS(seurat_obj, file = "./results/rds/Norm_all_samples.rds")
```

## Determine the dimensionality of the dataset  
how many components should we choose to include?  

**JackStrawPlot**  
*significant* PCs as those who have a strong enrichment of **low p-value**

* dashed line : uniform distribution
* solid curve above dash line : strong enrichments of features with low p-values  

*Cannot run Jackstraw plot on SCT data*  
```{r, eval=FALSE}
seurat_obj <- JackStraw(seurat_obj, num.replicate = 100, dims = pc)
jack.data <- seurat_obj 

seurat_obj <- ScoreJackStraw(seurat_obj, dims = 1:pc)

JackStrawPlot(seurat_obj, dims = 1:pc)
```

**ElbowPlot**  
ranking of principle components based on the percentage of variance explained by each one 

```{r, warning=FALSE, message=FALSE}
ElbowPlot(seurat_obj, ndims = pc)
```
