---
title: "QC for merged samples [Normalized]"
author: "ATIA Safiya"
date: "2023-06-26"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: lumen
---

# Environment  
## Load required packages  
```{r,  message=FALSE, warning=FALSE}
library(Seurat)
library(tidyverse)
library("SCINA")
library(ggplot2)
```

## Load seurat object (NORM)  
```{r}
seurat_obj <- readRDS(file = "Norm_merged_seurat.rds")

seurat_obj
table(seurat_obj$orig.ident)
```

# Metabo object  
```{r}
genes <- read.csv("all_metabo.tsv", sep = '\t', header = TRUE)
head(genes)
```

```{r}
features <- rownames(seurat_obj@assays$RNA@counts)
# keep genes names seeen un genes$Gene_name_version
matched_features <- features[features %in% genes$Gene_name_version]

counts <- GetAssayData(seurat_obj, assay = "RNA")
counts <- counts[(which(rownames(counts) %in% matched_features)),]
metabo_obj <- subset(seurat_obj, features = rownames(counts))
metabo_obj
```

## save seurat object (metabo)
```{r}
saveRDS(metabo_obj, file ="Norm_merged_seurat_METABO.rds")
```

# QC metrics  
All the percentages( MT, Ribosmoal, largest genes, expression) are the ones calculated compared to all genes, not only metabo ones.

```{r}
FeatureScatter(metabo_obj, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by ="orig.ident")
FeatureScatter(metabo_obj,feature1 = "nCount_RNA", feature2 = "percent.largest_gene", group.by ="orig.ident")

FeatureScatter(metabo_obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",group.by ="orig.ident") + geom_smooth(method = 'lm')
```
```{r, message=FALSE, warning=FALSE}
library(viridis)
metadata <- metabo_obj@meta.data
```

## Nb of reads  
```{r, message=FALSE}
VlnPlot(metabo_obj,features="nCount_RNA", group.by = "orig.ident", pt.size = 0)
RidgePlot(metabo_obj, features = "nCount_RNA", group.by = "orig.ident")

metadata %>% 
  ggplot(aes( x=nCount_RNA,y=orig.ident, fill= orig.ident)) + 
	geom_boxplot() + 
	theme_classic()+
	scale_fill_viridis(	option="magma",discrete = TRUE, alpha=0.6) +
    scale_x_log10()
```

## Nb of genes  
```{r, message=FALSE}
VlnPlot(metabo_obj,features="nFeature_RNA", group.by = "orig.ident", pt.size = 0)
RidgePlot(metabo_obj, features = "nFeature_RNA", group.by = "orig.ident")

metadata %>% 
  ggplot(aes( x=nFeature_RNA,y=orig.ident, fill= orig.ident)) + 
	geom_boxplot() + 
	theme_classic()+
	scale_fill_viridis(	option="magma",discrete = TRUE, alpha=0.6) +
    scale_x_log10()
    
```

```{r}
ggplot(qc.metrics) + 
   geom_histogram(aes(nFeature_RNA, 
                  color = "#173664"),
                  binwidth = 50) + 
  ggtitle("Distribution of nFeature_RNA ") + NoLegend()
```

## Gene expression  
We can pick the first 100 cells and look at the distributions of their expression values.

```{r, warning=FALSE}
apply(metabo_obj@assays$RNA@data,1,mean) -> gene.expression
sort(gene.expression, decreasing = TRUE) -> gene.expression

as.tibble(
  metabo_obj@assays$RNA@data[,1:100]
) %>%
  pivot_longer(
    cols=everything(),
    names_to="cell",
    values_to="expression"
  ) %>%
  ggplot(aes(x=expression, group=cell)) +
  geom_density() +
  coord_cartesian(ylim=c(0,0.6), xlim=c(0,3))
```

## MT genes  
```{r, message=FALSE}
#metabo_obj[["percent.mt"]] <- PercentageFeatureSet(metabo_obj, pattern = "^MT-")
VlnPlot(metabo_obj,features="percent.mt", group.by = "orig.ident", pt.size = 0)
RidgePlot(metabo_obj, features = "percent.mt", group.by = "orig.ident")
```

```{r}
ggplot(qc.metrics) + 
   geom_histogram(aes(percent.mt, 
                  color = "#173664"),
                  binwidth = 0.5) + 
  ggtitle("Distribution of Percentage Mitochondrion") + NoLegend()
```

## Ribosomal genes  
```{r, message=FALSE}
#PercentageFeatureSet(metabo_obj,pattern="^RP[LS]") -> metabo_obj[["percent.ribosomal"]] 
VlnPlot(metabo_obj,features="percent.ribosomal", group.by = "orig.ident", pt.size = 0)
RidgePlot(metabo_obj, features = "percent.ribosomal", group.by = "orig.ident")
```

```{r}
ggplot(qc.metrics) + 
   geom_histogram(aes(percent.ribosomal, 
                  color = "#173664"),
                  binwidth = 0.5) + 
  ggtitle("Distribution of Percentage Ribosomal") + NoLegend()
```

## Largest gene  
```{r, message=FALSE}
VlnPlot(metabo_obj,features="percent.largest_gene", group.by = "orig.ident", pt.size = 0)
RidgePlot(metabo_obj, features = "percent.largest_gene", group.by = "orig.ident")
```

```{r}
ggplot(qc.metrics) + 
   geom_histogram(aes(percent.largest_gene, 
                  color = "#173664"),
                  binwidth = 0.7) + 
  ggtitle("Distribution of Percentage Largest Gene") + NoLegend()
```

## Cell cycle scoring  
```{r, warning=FALSE}
seurat_obj <- CellCycleScoring(seurat_obj, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = TRUE)

seurat_obj@meta.data %>%
  group_by(orig.ident,Phase) %>%
  count() %>%
  group_by(orig.ident) %>%
  mutate(percent=100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=orig.ident,y=percent, fill=Phase)) +
  scale_fill_manual( values=c(G1='azure4',G2M='dodgerblue',S='tomato3' )) +
  geom_col() +
  ggtitle("Percentage of cell cycle phases per sample")

as_tibble(seurat_obj[[]]) %>%
  ggplot(aes(Phase, fill=Phase)) +
  scale_fill_manual( values=c(G1='azure4',G2M='dodgerblue',S='tomato3' ))+ geom_bar()


as_tibble(seurat_obj[[]]) %>%
  ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + 
  scale_color_manual( values=c(G1='azure4',G2M='dodgerblue',S='tomato3' )) +
  geom_point() +
  coord_cartesian(xlim=c(-0.15,0.15), ylim=c(-0.15,0.15))
```